extensions/amp-animation/0.1/keyframes-extractor.js

```diff
@@ -0,0 +1,162 @@
+ */
+export function extractKeyframes(rootNode, name) {
+  const styleSheets = rootNode.styleSheets;
+  if (!styleSheets) {
+    return null;
+  }
+  const win = (rootNode.ownerDocument || rootNode).defaultView;
+  // Go from the last to first since the last rule wins in CSS.
+  for (let i = styleSheets.length - 1; i >= 0; i--) {
+    const keyframes = scanStyle(
+        win, /** @type {!CSSStyleSheet} */ (styleSheets[i]), name);
+    if (keyframes) {
+      return keyframes;
+    }
+  }
+  return null;
+}
+
+
+/**
+ * @param {!Window} win
+ * @param {!CSSStyleSheet} styleSheet
+ * @param {string} name
+ * @return {?./web-animation-types.WebKeyframesDef}
+ */
+function scanStyle(win, styleSheet, name) {
+  // No rules, e.g. a font.
+  if (!styleSheet.cssRules) {
+    return null;
+  }
+
+  const styleNode = styleSheet.ownerNode;
+  if (!styleNode) {
+    return null;
+  }
+  // Exlcude AMP's own styles.
+  if (styleNode.hasAttribute('amp-boilerplate') ||
+      styleNode.hasAttribute('amp-runtime') ||
+      styleNode.hasAttribute('amp-extension')) {
+    return null;
+  }
+
+  return scanRules(win, styleSheet.cssRules, name);
+}
+
+
+/**
+ * @param {!Window} win
+ * @param {!CSSRuleList} rules
+ * @param {string} name
+ * @return {?./web-animation-types.WebKeyframesDef}
+ */
+function scanRules(win, rules, name) {
+  // Go backwards since in CSS the last one wins.
+  for (let i = rules.length - 1; i >= 0; i--) {
+    const rule = rules[i];
+    if (rule.type == /* CSSKeyframesRule */ 7) {
+      const keyframesRule = /** @type {!CSSKeyframesRuleDef} */ (rule);
+      if (rule.name == name && isEnabled(win, rule)) {
+        return buildKeyframes(keyframesRule);
+      }
+    } else if (rule.type == /* CSSMediaRule */ 4) {
+      // Go recursively inside. The media match will be checked only when
+      // the corresponding @keyframes have been found.
+      const found = scanRules(win, rule.cssRules, name);
+      if (found) {
+        return found;
+      }
+    }
+  }
+  return null;
+}
+
+
+/**
+ * @param {!Window} win
+ * @param {!CSSRule} rule
+ * @return {boolean}
+ */
+function isEnabled(win, rule) {
+  // Try rule itself.
+  if (rule.media && rule.media.mediaText) {
+    const enabled = win.matchMedia(rule.media.mediaText).matches;
+    if (!enabled) {
+      return false;
+    }
+  }
+
+  // Check the parents.
+  if (rule.parentRule) {
+    return isEnabled(win, rule.parentRule);
+  } else if (rule.parentStyleSheet &&
+      rule.parentStyleSheet.media &&
+      rule.parentStyleSheet.media.mediaText) {
+    return win.matchMedia(rule.parentStyleSheet.media.mediaText).matches;
+  }
+  return true;
+}
+
+
+/**
+ * @param {!CSSKeyframesRuleDef} keyframesRule
+ * @return {!./web-animation-types.WebKeyframesDef}
+ */
+function buildKeyframes(keyframesRule) {
+  const array = [];
+  for (let i = 0; i < keyframesRule.cssRules.length; i++) {
+    const keyframeRule = keyframesRule.cssRules[i];
+    const keyframe = {};
```

Comment:
I think @aghassemi meant map on all, e.g. `const array = keyframesRule.cssRules.map(rule => {})`. Unfortunately `cssRules` is not an array and I thought we decided to avoid simple `toArray` conversions just for better for-loops?